File: src/main/java/cspfj/constraint/semantic/SumLeq.java
package cspfj.constraint.semantic;

import cspfj.constraint.AbstractConstraint;
import cspfj.filter.RevisionHandler;
import cspfj.problem.Domain;
import cspfj.problem.Variable;

public final class SumLeq extends AbstractConstraint {

	private final int bound;

	public SumLeq(final int bound, final Variable... variables) {
		super(variables);
		this.bound = bound;
	}

	@Override
	public boolean check() {
		int sum = 0;
		for (int i = getArity(); --i >= 0;) {
			sum += getValue(i);
			if (sum > bound) {
				return false;
			}
		}

		return true;
	}

	@Override
	public float getEvaluation() {
		return getArity();
	}

	private boolean removeGt(final int value, final Domain dom) {
		final int lb = dom.lowest(value);
		if (lb >= 0) {
			if (dom.value(lb) != value) {
				return dom.removeFrom(lb) > 0;
			}
			return dom.removeFrom(lb + 1) > 0;
		}
		return false;
	}

	@Override
	public boolean revise(final RevisionHandler revisator, final int reviseCount) {
		int min = 0;
		for (int i = getArity(); --i >= 0;) {
			final Domain dom = getVariable(i).getDomain();
			min += dom.value(dom.first());
		}
		final int newBound = bound - min;
		if (newBound < 0) {
			return false;
		}

		int max = 0;
		for (int i = getArity(); --i >= 0;) {
			final Domain dom = getVariable(i).getDomain();
			if (removeGt(newBound + dom.value(dom.first()), dom)) {
				if (dom.size() <= 0) {
					return false;
				}
				revisator.revised(this, getVariable(i));
			}
			max += dom.value(dom.last());
		}

		if (max <= bound) {
			entail();
		}

		return true;
	}

	@Override
	public String toString() {
		final StringBuilder stb = new StringBuilder();
		stb.append(getVariable(0));
		for (int i = 1; i < getArity(); i++) {
			stb.append(" + ").append(getVariable(i));
		}
		stb.append(" <= ").append(bound);
		return stb.toString();
	}
}

File: src/main/scala/cspfj/constraint/semantic/SumLeq.scala
package cspfj.constraint.semantic;

import cspfj.constraint.AbstractConstraint
import cspfj.filter.RevisionHandler
import cspfj.problem.Domain
import cspfj.problem.Variable
import scala.annotation.tailrec
import cspfj.constraint.SimpleRemovals
import cspfj.util.Loggable

final class SumLeq(val bound: Int, scope: Array[Variable]) extends AbstractConstraint(scope)
with SimpleRemovals with Loggable {

  def check: Boolean = {

    @tailrec
    def c(i: Int, bound: Int): Boolean = {
      if (i >= arity) true else {
        val nb = bound - value(i)
        if (nb < 0) false else c(i + 1, nb)
      }
    }

    c(0, bound)
  }

  def getEvaluation = arity

  private def removeGt(value: Int, dom: Domain) = {
    val lb = dom.closestGeq(value)
    if (lb >= 0) {
      if (dom.value(lb) != value) {
        dom.removeFrom(lb) > 0;
      } else {
        dom.removeFrom(lb + 1) > 0;
      }
    } else false
  }

  def revise(revisator: RevisionHandler, reviseCount: Int): Boolean = {
    //info(this.toString)
    val newBound = bound - scope.iterator.map(_.dom.firstValue).sum
    if (newBound < 0) {
      return false;
    }

    var max = 0;
    for (v <- scope) {
      val dom = v.dom
      if (removeGt(newBound + dom.firstValue, dom)) {
        if (dom.size == 0) {
          return false;
        }
        revisator.revised(this, v);
      }
      max += dom.lastValue
    }

    if (max <= bound) {
      entail();
    }

    true;
  }

  override def toString = scope.mkString(" + ") + " <= " + bound
}
